src/workerd/io/worker.c++

```diff
@@ -1060,6 +1113,46 @@ Worker::Isolate::Isolate(kj::Own<ApiIsolate> apiIsolateParam,
       }
     }
   });
+
+  // The PromiseCrossContextCallback is used to allow cross-IoContext promise following.
+  // When the IoContext::Scope is entered, we set the "promise context tag" associated
+  // with the IoContext on the Isolate that is locked. Any Promise that is created within
+  // that scope will be tagged with the same promise context tag. When an attempt to
+  // follow a promise occurs (e.g. either using Promise.prototype.then() or await, etc)
+  // our patched v8 logic will check to see if the followed promise's tag matches the
+  // current Isolate tag. If they do not, then v8 will invoke this callback. The promise
+  // here is the promise that belongs to a different IoContext.
+  lock->v8Isolate->SetPromiseCrossContextCallback([](v8::Local<v8::Context> context,
+                                                     v8::Local<v8::Promise> promise,
+                                                     v8::Local<v8::Object> tag) ->
+                                                       v8::MaybeLocal<v8::Promise> {
+    if (!IoContext::hasCurrent()) return promise;
+    auto isolate = context->GetIsolate();
+    auto& js = jsg::Lock::from(isolate);
+    auto promiseRef = js.v8Ref(promise);
+    // The CrossThreadPromiseWaiter is a help that is attached to the promise. It is
+    // allows multiple cross-thread/cross-request waiters to be attached to the promise,
+    // all of which will be fulfilled when promise is resolved.
+    auto& waiter = getCrossThreadPromiseWaiter(js, promiseRef);
```

Comment:
If `promise` is already resolved, then there's no need to set up a cross-thread wait. You can safely just return it. Can we check `promise->State()` and optimize here?

Reply:
If `promise` is already resolved this callback wouldn't have been called.

Reply:
Ah, great.